storage/innobase/row/row0ins.cc

```diff
@@ -2703,22 +2703,43 @@ row_ins_clust_index_entry_low(
 
 	DBUG_EXECUTE_IF("row_ins_row_level", goto skip_bulk_insert;);
 
-	if (!(flags & BTR_NO_UNDO_LOG_FLAG)
-	    && page_is_empty(block->page.frame)
-	    && !entry->is_metadata() && !trx->duplicates
-	    && !trx->check_unique_secondary && !trx->check_foreigns
-	    && !trx->dict_operation
-	    && block->page.id().page_no() == index->page
-	    && !index->table->skip_alter_undo
+	if (((!(flags & BTR_NO_UNDO_LOG_FLAG)
+	      && page_is_empty(block->page.frame)
+	      && !entry->is_metadata() && !trx->duplicates
+	      && !trx->check_unique_secondary && !trx->check_foreigns
+	      && !trx->dict_operation
+	      && !index->table->is_active_ddl()
+	      && block->page.id().page_no() == index->page)
+	     || (flags == (BTR_NO_UNDO_LOG_FLAG | BTR_NO_LOCKING_FLAG)
+		 && !index->table->is_temporary()
+		 && index->table->foreign_set.empty()))
 	    && !index->table->n_rec_locks
```

Comment:
In the `flags == BTR_NO_UNDO_LOG_FLAG | BTR_NO_LOCKING_FLAG` case, I think that it would be better to assert that some of the conditions that are being checked in the `!(flags & BTR_NO_UNDO_LOG_FLAG)` hold. Maybe this could be rewritten for clarity as a nested `if`:
```c++
if (!(flags & BTR_NO_UNDO_LOG_FLAG) && … && !index->table->has_spatial_index()
    && !(dict_table_is_partition(index->table || thd_sql_command(…) == SQLCOM_INSERT)) {
	if (!index->table->n_rec_locks && !index->table->versioned() && !index->table->skip_alter_undo) {
			trx->bulk_insert = true;
	}
} else if (flags == (BTR_NO_UNDO_LOG_FLAG | BTR_NO_LOCKING_FLAG)
	   && …) {
	ut_ad(index->table->skip_alter_undo);
	ut_ad(!entry->is_metadata());
	ut_ad(index->table->is_active_ddl());
	if (innodb_alter_copy_bulk) {
		ut_ad(page_is_empty(block->page.frame));
		trx->bulk_insert = true;
	}
}
```
Would this make sense? Add  as many assertions as you think should hold. Do we really need the `index->table->is_active_ddl()` condition for the `!(flags & BTR_NO_UNDO_LOG_FLAG)` case?